testsuite/tests/lib-random/chi2.ml

```diff
@@ -95,6 +111,175 @@ let _ =
+      (let p = 4194319 in (* prime > 2^22 and < 2^23 *)
+       let min_ = min_int in
+       let max_ = min_ + (256 * p) - 1 in
+       fun () -> (Random.int_in_range ~min:min_ ~max:max_ - min_) / p);
+  if Sys.int_size >= 32 then begin
+    let min_int32 = Int64.to_int (-0x8000_0000L) in
+    let max_int32 = Int64.to_int 0x7FFF_FFFFL in
+    test "Random.int_in_range, full int32 range (bits 0-7)"
+        (fun () -> Random.int_in_range ~min:min_int32 ~max:max_int32);
+    test "Random.int_in_range, full int32 range (bits 24-31)"
+        (fun () -> (Random.int_in_range ~min:min_int32 ~max:max_int32) lsr 24);
+
+    test "Random.int_in_range, range of length 2^31 < 256*p < 2^32 (bits 0-7)"
+        (let p = 8388617 in (* prime > 2^23 and < 2^24 *)
+         let min_ = min_int in
+         let max_ = min_ + (256 * p) - 1 in
+         fun () -> (Random.int_in_range ~min:min_ ~max:max_ - min_) / p);
+  end;
+  if Sys.int_size >= 63 then begin
+    let min_ = Int64.to_int (-1844674407370955197L) in
+    let max_ = min_ + Int64.to_int 0x1FFF_FFFF_FFFF_FFFFL in
+    test "Random.int_in_range, range of length 2^61 (bits 0-7)"
+         (fun () -> Random.int_in_range ~min:min_ ~max:max_ - min_);
+    test "Random.int_in_range, range of length 2^61 (bits 30-37)"
+         (fun () -> (Random.int_in_range ~min:min_ ~max:max_ - min_) lsr 30);
+    test "Random.int_in_range, range of length 2^61 (bits 53-60)"
+         (fun () -> (Random.int_in_range ~min:min_ ~max:max_ - min_) lsr 53);
+    let min_ = Int64.to_int (-1844674407370955197L) in
+    let max_ = min_ + Int64.to_int 0x3FFF_FFFF_FFFF_FFFFL in
+    test "Random.int_in_range, range of length 2^62 (bits 0-7)"
+         (fun () -> Random.int_in_range ~min:min_ ~max:max_ - min_);
+    test "Random.int_in_range, range of length 2^62 (bits 30-37)"
+         (fun () -> (Random.int_in_range ~min:min_ ~max:max_ - min_) lsr 30);
+    test "Random.int_in_range, range of length 2^62 (bits 54-61)"
+         (fun () -> (Random.int_in_range ~min:min_ ~max:max_ - min_) lsr 54);
+    test "Random.int_in_range, range of length 2^61 < 256*p < 2^62 (bits 0-7)"
+        (let p = Int64.to_int 9007199254740997L in (* prime > 2^53 and < 2^54 *)
+         let min_ = min_int in
+         let max_ = min_ + (256 * p) - 1 in
+         fun () -> (Random.int_in_range ~min:min_ ~max:max_ - min_) / p);
+    test "Random.int_in_range, range of length 256*p > 2^62 (bits 0-7)"
+        (let p = Int64.to_int 18014398509482143L in (*prime > 2^54 and < 2^55 *)
+         let min_ = min_int in
+         let max_ = min_ + (256 * p) - 1 in
+         fun () -> (Random.int_in_range ~min:min_ ~max:max_ - min_) / p);
+  end;
+  test "Random.int_in_range, full int range (bits 0-7)"
+       (fun () -> Random.int_in_range ~min:min_int ~max:max_int);
+  test "Random.int_in_range, full int range (bits 30-37)"
+       (fun () -> (Random.int_in_range ~min:min_int ~max:max_int) lsr 30);
```

Comment:
That test is what fails on i386. Obviously when `int` is 31 bits, sampling bits 30--37 is not uniform. This test should be moved in the conditional block `if Sys.int_size >= 63` (are 31, 32, 63 the only integer sizes we care about?).